ruby_programming/testing_with_rspec/introduction_to_rspec.md

```diff
@@ -0,0 +1,236 @@
+    it "returns the sum of two numbers" do
+      calculator = Calculator.new
+      expect(calculator.add(5, 2)).to eql(7)
+    end
+  end
+end
+```
+
+Let's go line by line.
+
+First, `describe` is an RSpec keyword that defines an "Example Group", or a collection of tests. It takes a class or a string as an argument and is passed a block (`do/end`). `describe` blocks can be nested, such as on the second line of our test above. When describing a class, the following syntax is also valid:
+
+```ruby
+#spec/calculator_spec.rb
+
+describe Calculator do
+  #...
+end
+
+```
+
+The `it` keyword defines an individual example (aka, test). `it` takes a string argument and is also passed a block. This block is where our expectations of a method are expressed. In this particular case, when we pass 5 and 2 to the `#add` method, we expect it to return 7. This is concisely expressed in our expectation clause, which uses one of RSpec's equality [matchers](https://relishapp.com/rspec/rspec-expectations/v/3-7/docs/built-in-matchers/equality-matchers), `eql`:
+
+```
+expect(calculator.add(5, 2)).to eql(7)
+```
+
+Simple, isn't it? One more time, from the top:
+1. `describe` the class
+2. `describe` the method example group. Conventionally, the string argument for instance methods are written as "#method", while string arguments for class methods are written as ".method".
+3. Write your test case/example with `it`. 
+4. Write your expectation using `expect`.  The `expect` method is also chained with `.to` for positive expectations, or `.to_not`/`.not_to` for negative expectations. We prefer `.not_to`. Also, limit one expect clause per test case.
+
+### Passing code
+
+Let's move on. Run `rspec` from the directory root, and watch the output.
+
+```bash
+An error occurred while loading ./spec/calculator_spec.rb.
+Failure/Error:
+  describe Calculator do
+    describe "#add" do
+      it "returns the sum of two numbers" do
+        calculator = Calculator.new
+        expect(calculator.add(5, 2)).to eql(7)
+      end
+    end
+  end
+
+NameError:
+  uninitialized constant Calculator
+# ./spec/calculator_spec.rb:1:in `<top (required)>'
+No examples found.
+
+
+Finished in 0.0004 seconds (files took 0.16461 seconds to load)
+0 examples, 0 failures, 1 error occurred outside of examples
+```
+
+So our first test returned an error. This is unsurprising. `NameError` is essentially telling us that RSpec looked for a `Calculator` class, but couldn't find one. So let's create it. From your project root, create a `lib/` folder, and inside, `calculator.rb` with your class. We'll also go ahead and begin the `#add` method, otherwise RSpec will give us a similar error as the previous one when it looks for it:
+
+```ruby
+#lib/calculator.rb
+
+class Calculator
+  def add(a,b)
+  end
+end
+```
+
+Finally, we must also tell the spec where the `Calculator` class is being defined. This is easily done with `require`:
+
+```ruby
+#spec/calculator_spec.rb
+require './lib/calculator'
```

Comment:
Add an `# => Added this` comment to the `require` line to make it stand out. SEE: Line 188 for example

Reply:
:+1: